include/mfmg/dealii/lanczos_lanczos.template.hpp

```diff
@@ -0,0 +1,398 @@
+
+  calc_evecs_();
+
+  if (verbosity_ > 0) {
+    std::cout << std::endl;
+  }
+}
+
+//-----------------------------------------------------------------------------
+/// \brief Lanczos solver: calculate eigenpairs from tridiag of lanczos coeffs
+
+template<typename Op_t>
+void Lanczos<Op_t>::calc_tridiag_epairs_(int it, Scalars_t& t_maindiag,
+                                         Scalars_t& t_offdiag) {
+  assert(it >= 1);
+  assert(it <= this->maxit_);
+  assert(t_maindiag.size() == it);
+  assert(t_offdiag.size() == it-1);
+
+  dim_tridiag_ = it;
+  const int k = dim_tridiag_;
+
+  // Set up matrix
+
+  // TODO: an implementation of this with fewerpointers and more encapsulation.
+
+  auto matrix = new double[k*k];
+
+  for (int i=0; i<k*k; ++i) {
+    matrix[i] = 0;
+  }
+
+  // Copy tridiag matrix to full matrix.
+
+  matrix[0] = (double)t_maindiag[0];
+  for (int i=1; i<k; ++i) {
+    matrix[i + k*i] = (double)t_maindiag[i];
+    matrix[i + k*(i-1)] = (double)t_offdiag[i-1];
+    matrix[i-1 + k*i] = (double)t_offdiag[i-1];
+  }
+
+  // LAPACK eigenvalue/vector solve.
+
+  // NOTE: this part can be replaced if desired with some platform-specific
+  // library.
+  // NOTE: for accuracy we are using double here rather than Scalar_t;
+  // may not be needed.
+  // ISSUE: LAPACK has other solvers that might be more efficient here.
+
+  auto t_evals_r = new double[k];
+  auto t_evals_i = new double[k];
+  auto t_evecs = new double[k*k];
+
+  // Do all in double, regardless of Scalar_t (for now)
+  const lapack_int info = LAPACKE_dgeev(LAPACK_COL_MAJOR, 'N', 'V',
+      k, matrix, k, t_evals_r, t_evals_i, NULL, k, t_evecs, k);
+
+  // Sort into ascending order
+
+  std::vector<int> sort_index(k);
+  for (int i=0; i<k; ++i) {
+    sort_index[i] = i;
+  }
+
+  std::sort(sort_index.begin(), sort_index.end(),
+    [&](int i, int j) { return t_evals_r[i] <
+                               t_evals_r[j]; });
+
+  // Output requested eigenvalues.
+
+  if (verbosity_ > 0) {
+    std::cout.width(4);
+    std::cout << "It " << k;
+    std::cout.precision(4);
+    std::cout << "   evals ";
+    for (int i=0; i<k && i<num_requested_; ++i) {
+      std::cout << " " << std::fixed << t_evals_r[sort_index[i]];
+    }
+    if (dim_tridiag_ < num_requested_) {
+      std::cout << "\n";
+    }
+  }
+
+  // Save results.
+
+  if (k >= num_requested_) {
+    evals_.resize(num_requested_);
+    evecs_tridiag_.resize(k * num_requested_);
+
+    for (int i=0; i<num_requested_; ++i) {
+      const int si = sort_index[i];
+      evals_[i] = (Scalar_t)t_evals_r[si];
+      double sum = 0;
+      for (int j=0; j<k; ++j) {
+        sum += t_evecs[j+k*si] * t_evecs[j+k*si];
+      }
+      const double norm = sqrt(sum);
+      for (int j=0; j<k; ++j) {
+        evecs_tridiag_[j+k*i] = (Scalar_t)(t_evecs[j+k*si] / norm);
+      }
+    } // for i
+  } // if
+
+  // Terminate
+
+  delete matrix;
+  delete t_evals_r;
+  delete t_evals_i;
+  delete t_evecs;
+
+#if 0
```

Comment:
What is this code for?

Reply:
@wdj

Reply:
Code under the #if 0 - just a comment, a mental reminder of the call signature of those possibly relevant lapack routines.